src/Entity/OgRole.php

```diff
@@ -165,4 +200,123 @@ public function save() {
     parent::save();
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function set($property_name, $value) {
+    // Prevent the ID, role type, entity type or bundle to be changed for any of
+    // the default roles. These default roles are required and shouldn't be
+    // tampered with.
+    $is_locked_property = in_array($property_name, [
+      'id',
+      'role_type',
+      'group_type',
+      'group_bundle',
+    ]);
+    $is_default_role = $this->getRoleType() !== self::ROLE_TYPE_STANDARD;
+    if ($is_locked_property && $is_default_role && !$this->isNew()) {
+      throw new OgRoleRequiredException("The $property_name of the default roles 'non-member' and 'member' cannot be changed.");
+    }
+    return parent::set($property_name, $value);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function delete() {
+    // The default roles are required. Prevent them from being deleted for as
+    // long as the group still exists.
+    if (in_array($this->id(), [self::ANONYMOUS, self::AUTHENTICATED]) && $this->groupManager()->isGroup($this->getGroupType(), $this->getGroupBundle())) {
+      throw new OgRoleRequiredException('The default roles "non-member" and "member" cannot be deleted.');
+    }
+    parent::delete();
+  }
+
+  /**
+   * Returns default properties for each of the standard role names.
+   *
+   * @param string $role_name
+   *   A role name, one of OgRoleInterface::ANONYMOUS,
+   *   OgRoleInterface::AUTHENTICATED, OgRoleInterface::ADMINISTRATOR.
+   *
+   * @return array
+   *   An array of default properties, to pass to OgRole::create().
+   */
+  public static function getDefaultProperties($role_name) {
```

Comment:
I think it's better to have the getInfo logic separated from the create -
like this we can get the default roles info without creating them.

However, sure - create() can call this method for us and do the heavy
lifting / error throwing for invalid data
On May 18, 2016 5:01 PM, "Pieter Frenssen" <notifications@github.com> wrote:

> In src/Entity/OgRole.php
> <https://github.com/amitaibu/og/pull/192#discussion_r63707656>:
>
> > +      throw new OgRoleRequiredException('The default roles "non-member" and "member" cannot be deleted.');
> > +    }
> > +    parent::delete();
> > +  }
> > +
> > +  /**
> > +   * Returns default properties for each of the standard role names.
> > +   *
> > +   * @param string $role_name
> > +   *   A role name, one of OgRoleInterface::ANONYMOUS,
> > +   *   OgRoleInterface::AUTHENTICATED, OgRoleInterface::ADMINISTRATOR.
> > +   *
> > +   * @return array
> > +   *   An array of default properties, to pass to OgRole::create().
> > +   */
> > +  public static function getDefaultProperties($role_name) {
>
> I have changed it now to getProperties() but it feels wrong somehow.
> Since this is on the OgRole class it now seems to imply that it will
> return the properties of the role that is represented by the class.
>
> I'm now thinking that having this static method here is not useful after
> all. When creating a role programmatically we would always have to call
> this method to get a list of default properties, then pass it on to
> OgRole::create(). It would be much more interesting if this was built-in
> to OgRole::create().
>
> OK if I change this to a protected method that will be called from
> OgRole::create()?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/amitaibu/og/pull/192/files/4baf10c048a58f032ba6963f453be4ac4a4c1b1a#r63707656>
>